### PR TITLE
Fix SoftmaxOps

### PR DESCRIPTION
**Pull request by jerryzh168**
Summary:
We might see the pattern
```
if (scale_.numel() != N) {
   scale_->Resize(N);
   // set initial value for scale_
}

// In class:
Tensor scale_{CPU};
```
before in the code, where `scale_` is a member variable of Type `caffe2::Tensor`
This pattern actually has two functions, `scale_` is partially initialized with device type but not size, this call will
initialize Tensor with the correct size, or if `scale_` is already initialized with size, it will check whether the size
matches a runtime value `N` and if not it will Resize. To rewrite this we'll do the following:
```
if (!scale_.has_value()) {
  scale_ = caffe2::empty({N}, at::dtype<float>().device(CPU));
  // set initial value for scale_
} else if (scale_->numel() != N) {
  scale_->Resize(N);
  // set initial value for scale_
}

// in class:
c10::optional<Tensor> scale_;
```
There are some variants, if `scale_` is resized to a constant size, we can call `ReinitialieTensor` instead
```
if (scale_.numel() != 1) {
  scale_->Resize(1);
}
```
-->
```
ReinitializeTensor(&scale_, {1}, at::dtype<float>().device(CPU));
```

Normal Resize will be refactored directly into ReinitializeTensor:
```
scale_->Resize(N);
```
-->
```
ReinitializeTensor(&scale_, {N}, at::dtype<float>().device(CPU));
```

Differential Revision: D13667883

### CODE REVIEW

File: caffe2/operators/softmax_op.cc [hunk 1]
```diff
@@ -15,28 +15,36 @@ bool SoftmaxOp<float, CPUContext>::RunOnDevice() {
   float* Ydata = Y->template mutable_data<float>();
   // ReinitializeTensor itself has the effect of caching, so there is no need to check for numel of Tensor
   // First, get scales
```
Comment (ezyang) [inline caffe2/operators/softmax_op.cc]: This comment is out of date now

**Comment by ezyang** [comment]
Sorry, dumb question. I thought ReinitializeTensor didn't resize if the size matched up with the old version. So what's the difference between the new code you've written here, versus ReinitializeTensor?

**Comment by jerryzh168** [comment]
@ezyang https://fb.quip.com/IJ3ZA8ye5RwF, I'll update the summary.

**Comment by ezyang** [comment]
You added the following text to the description:

> Also, note that we use ReinitializeTensor only because it is easier to write codemod code for it, the recommended way should be to use c10::optional and caffe2::empty since that is more explicit and intuitive.

and you updated the example to be:

```
if (!scale_.has_value()) {
  scale_ = caffe2::empty({N}, at::dtype<float>().device(CPU));
  // set initial value for scale_
} else if (scale_->numel() != N) {
  scale_->Resize(N);
  // set initial value for scale_
}

// in class:
c10::optional<Tensor> scale_;
```

Seeing this example now, I don't actually think optional is appropriate for this idiom. The problem is that, elsewhere in the code, you are probably going to need to use `scale_`. And consequently, you are probably going to write `*scale_` or `scale_->data()` to get at the data inside scale. But this is sort of missing the point of having an optional, which is to force you to think about when you use a value that might be nullopt, and handle both cases appropriately. If we are going to enforce some invariant purely by data flow (as is the case here), might as well use undefined tensor and cut the line noise...

**Comment by jerryzh168** [comment]
@ezyang the reason for optional here is we need to check to see whether `scale_` is initialized or not in order to decide what to do, as shown in the example, we'll fill the data of `scale_` only when it is uninitialized or when the size does not match, and I think the rest of the code can assume it is initialized after this step?
In other words, if we have a function on Tensor that tells us whether it is undefined, that would work too.

**Comment by jerryzh168** [comment]
I think this is more explicit than `ReinitailizeTensor` since it exposes the logic of when to initialize and propagate the value of data in the Tensor.

**Comment by ezyang** [comment]
Yes there is such a function, it is `x.defined()`. It doesn't exist on `caffe2::Tensor` at the moment but that can be fixed.

**Comment by jerryzh168** [comment]
@ezyang OK, I can add that

File: caffe2/operators/spatial_softmax_with_loss_op.h [hunk 2]
```diff
@@ -28,10 +28,10 @@ class SpatialSoftmaxWithLossOp final : public Operator<Context> {
...
   StorageOrder order_;
 
-  Tensor losses_; // Per example loss
-  Tensor rowmax_{Context::GetDeviceType()}; // per example row max
-  Tensor weights_; // unignored weights
-  Tensor sum_multiplier_; // Vector of ones for summing via dot prod
+  c10::optional<Tensor> losses_; // Per example loss
+  c10::optional<Tensor> rowmax_; // per example row max
+  c10::optional<Tensor> weights_; // unignored weights
+  c10::optional<Tensor> sum_multiplier_; // Vector of ones for summing via dot prod
```
Comment (ezyang) [inline caffe2/operators/spatial_softmax_with_loss_op.h]: ?

**Pull request closed without merging**